Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -662,161 +664,195 @@ WarpXOpenPMDPlot::DumpToFile (ParticleContainer* pc,
+        currSpecies.setAttribute("currentDeposition", []() {
+            switch (WarpX::current_deposition_algo) {
+                case CurrentDepositionAlgo::Esirkepov :
+                    return "Esirkepov";
+                case CurrentDepositionAlgo::Vay :
+                    return "Vay";
+                default:
+                    return "directMorseNielson";
+            }
+        }());
+    }
 
-  //
-  // define positions & offsets
-  //
-  const unsigned long long NewParticleVectorSize = counter.GetTotalNumParticles() + ParticleFlushOffset;
-  m_doParticleSetUp = false;
-  if (counter.GetTotalNumParticles() > 0 and ParticleFlushOffset == 0) {
-      // This will trigger meta-data flush for particles the first-time non-zero number of particles are flushed.
-      m_doParticleSetUp = true;
-  }
-  SetupPos(currSpecies, NewParticleVectorSize, charge, mass, isBTD);
-  SetupRealProperties(pc, currSpecies, write_real_comp, real_comp_names, write_int_comp, int_comp_names, NewParticleVectorSize, isBTD);
-  // open files from all processors, in case some will not contribute below
-  m_Series->flush();
-  for (auto currentLevel = 0; currentLevel <= pc->finestLevel(); currentLevel++)
-    {
-      uint64_t offset = static_cast<uint64_t>( counter.m_ParticleOffsetAtRank[currentLevel] );
-      // For BTD, the offset include the number of particles already flushed
-      if (isBTD) offset += ParticleFlushOffset;
-      for (ParticleIter pti(*pc, currentLevel); pti.isValid(); ++pti) {
-         auto const numParticleOnTile = pti.numParticles();
-         uint64_t const numParticleOnTile64 = static_cast<uint64_t>( numParticleOnTile );
-
-         // Do not call storeChunk() with zero-sized particle tiles:
-         //   https://github.com/openPMD/openPMD-api/issues/1147
-         if (numParticleOnTile == 0) continue;
-
-         // get position and particle ID from aos
-         // note: this implementation iterates the AoS 4x...
-         // if we flush late as we do now, we can also copy out the data in one go
-         const auto& aos = pti.GetArrayOfStructs();  // size =  numParticlesOnTile
-         {
-           // Save positions
-           auto const positionComponents = detail::getParticlePositionComponentLabels();
+    // define positions & offset structure
+    const unsigned long long NewParticleVectorSize = total_particles + ParticleFlushOffset;
+    // we will set up empty particles unless it's BTD, where we might add some in a following buffer dump
+    //   during this setup, we mark some particle properties as constant and potentially zero-sized
+    m_doParticleSetUp = true;
+    if (isBTD)
+        m_doParticleSetUp = is_first_flush_with_particles || is_last_flush_and_never_particles;
+
+    // this setup stage also implicitly calls "makeEmpty" if needed (i.e., is_last_flush_and_never_particles)
+    SetupPos(currSpecies, NewParticleVectorSize, charge, mass, isBTD);
```

Comment:
When m_doParticleSetup is false, should it still go to Setup*() functions?

Reply:
yes, there is handling inside them to skip parts if `m_doParticleSetUp` is set.

Reply:
Good question, we can check if we can skip that altogether and delay it in BTD until we fulfill `m_doParticleSetUp`.

Reply:
Ah, the challenge here is that we need to call this setup multiple times for BTD, since we use it to resize the output as we add more particles to the same step in subsequent calls.

Reply:
when there is no particle, nothing to adjust?

Reply:
I have to adjust a bit more next week for BTD. Have to move out the declaration of constant records so they only happen once

Reply:
Done :)